content/articles/writing-my-first-crate.md

```diff
@@ -0,0 +1,156 @@
+
+
+## Enter jsonwebtoken
+I was implementing JWT in a Go project using the [jwt-go](https://github.com/dgrijalva/jwt-go) library when I realised a JWT library was something simple enough that I could try in Rust.
+
+We can see there are already [crates](https://crates.io/search?q=jwt) for that available but [one](https://github.com/GildedHonour/frank_jwt) only allows string types and the other put too much important on the registered claims to my liking.
+
+I didn't implement the standard to the letter as I wanted to keep it simple for now and not support some part of it (having `none` a valid value for the alg header makes no sense to me therefore is not supported).
+
+
+Compared to the Go library (which is quite good!), using Rust allows leveraging generics to have type-safe claims.
+For example, the claims in jwt-go are stored in a `map[string]interface{}` which is the only way to do so and therefore need type assertion when decoding claims to access its true type. The next version (v3 at the time of writing) seem to allow passing a struct though.
+
+Here's how jsonwebtoken is used in practice, taken from the example in the crate:
+
+```rust
+extern crate jsonwebtoken as jwt;
+extern crate rustc_serialize;
+
+use jwt::{
+    Algorithm,
+    encode,
+    decode
+};
+
+// Deriving RustcEncodable and RustcDecodable make Claims satisfy the jwt::Part trait
+#[derive(Debug, RustcEncodable, RustcDecodable)]
+struct Claims {
+    sub: String,
+    company: String
+}
+
+fn main() {
+    let my_claims = Claims {
+        sub: "b@b.com".to_owned(),
+        company: "ACME".to_owned()
+    };
+    let key = "secret";
+
+    let token = match encode::<Claims>(my_claims, key.to_owned(), Algorithm::HS256) {
+        Ok(t) => t,
+        Err(_) => panic!()
+    };
+    // or let token = try!(encode::<Claims>(my_claims, key.to_owned(), Algorithm::HS256));
+
+    // And then decode it back into a Claims object
+    let claims = match decode::<Claims>(token.to_owned(), key.to_owned(), Algorithm::HS256) {
+        Ok(c) => c,
+        Err(err) => match err {
+            Error::InvalidToken => panic!(), // Example on how to handle a specific error
+            _ => panic!()
+        }
+    };
+    // or let claims = try!(decode::<Claims>(token.to_owned(), key.to_owned(), Algorithm::HS256));
+}
+```
+
+On the other hand, decoding the token gives you an instance of the `Claims` struct declared in the function and thanks to the `Result` type, we are certain that if we are in the `Ok()` branch, the variable contains valid data.
+
+How cool is that. And as long as the struct we use satisfy the [`Part` trait](https://github.com/Keats/rust-jwt/blob/6ae77c0b068328c47febe4169d6d28c0c66ba101/src/lib.rs#L29-L47), we are free to use anything we want. I'm guessing other languages with generics implement similar things for their JWT libraries but compared to the Go one or the [python one](https://github.com/jpadilla/pyjwt) (now you can see where I got the API from really).
```

Comment:
- `How cool is that` could be nicer to read with a `?` or a `?!`
- `we are free to use anything we want` if you mean every struct, I'd suggest `we can use any struct we want as long as it implements the Part trait. It might be worth to touch on how you make a struct implement a trait and/or what it means. If you're going for the rust vs Go thing maybe even compare to interfaces
- `but compared to the Go one or the python one [...]` feels like an unfinished sentence. after `compared to ...` I expected a `the rust implementation is ... `